File: app/src/main/java/com/vkpapps/thunder/ui/adapter/AudioAdapter.java
package com.vkpapps.thunder.ui.adapter;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.RadioButton;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.appcompat.widget.AppCompatImageView;
import androidx.recyclerview.widget.RecyclerView;

import com.google.android.gms.ads.AdView;
import com.vkpapps.thunder.R;
import com.vkpapps.thunder.model.AudioInfo;
import com.vkpapps.thunder.utils.AdsUtils;
import com.vkpapps.thunder.utils.MyThumbnailUtils;
import com.vkpapps.thunder.utils.StorageManager;

import java.io.File;
import java.util.List;

/**
 * @author VIJAY PATIDAR
 */
public class AudioAdapter extends RecyclerView.Adapter<AudioAdapter.AudioViewHolder> {
    private List<AudioInfo> audioInfos = null;
    private OnAudioSelectedListener onAudioSelectedListener;
    private File thumbnails;
    private LayoutInflater inflater;
    private MyThumbnailUtils myThumbnailUtils = MyThumbnailUtils.INSTANCE;

    public AudioAdapter(@NonNull OnAudioSelectedListener onAudioSelectedListener, @NonNull Context context) {
        this.onAudioSelectedListener = onAudioSelectedListener;
        thumbnails = new StorageManager(context).getThumbnails();
        inflater = LayoutInflater.from(context);
    }

    @NonNull
    @Override
    public AudioViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View inflate;
        if (viewType == 1) {
            inflate = inflater.inflate(R.layout.audio_list_item, parent, false);
        } else {
            inflate = inflater.inflate(R.layout.local_list_item_ad_view, parent, false);
        }
        return new AudioViewHolder(inflate);
    }

    @Override
    public int getItemViewType(int position) {
        return audioInfos.get(position) == null ? 0 : 1;
    }

    @Override
    public void onBindViewHolder(@NonNull final AudioViewHolder holder, int position) {
        final AudioInfo audioinfo = audioInfos.get(position);
        if (audioinfo == null) {
            AdView adView = (AdView) holder.itemView;
            AdsUtils.INSTANCE.getAdRequest(adView);
        } else {
            holder.audioTitle.setText(audioinfo.getName());
            holder.btnSelect.setChecked(audioinfo.isSelected());
            holder.btnSelect.setOnClickListener((v) -> {
                audioinfo.setSelected(!audioinfo.isSelected());
                holder.btnSelect.setChecked(audioinfo.isSelected());
                if (audioinfo.isSelected()) {
                    onAudioSelectedListener.onAudioSelected(audioinfo);
                } else {
                    onAudioSelectedListener.onAudioDeselected(audioinfo);
                }
            });

            File file = new File(thumbnails, audioinfo.getId());
            myThumbnailUtils.loadAudioThumbnail(file, audioinfo.getPath(), holder.audioIcon);
            holder.audioIcon.setAdjustViewBounds(true);
        }
    }

    @Override
    public int getItemCount() {
        return (audioInfos == null) ? 0 : audioInfos.size();
    }

    static class AudioViewHolder extends RecyclerView.ViewHolder {

        TextView audioTitle;
        AppCompatImageView audioIcon;
        RadioButton btnSelect;
        AudioViewHolder(@NonNull View itemView) {
            super(itemView);
            audioIcon = itemView.findViewById(R.id.audio_icon);
            audioTitle = itemView.findViewById(R.id.audio_title);
            btnSelect = itemView.findViewById(R.id.btnSelect);
        }

    }

    public interface OnAudioSelectedListener {
        void onAudioSelected(AudioInfo audioMode);

        void onAudioDeselected(AudioInfo audioinfo);
    }

    public void setAudioInfos(List<AudioInfo> audioInfos) {
        this.audioInfos = audioInfos;
        notifyDataSetChanged();
    }
}

File: app/src/main/java/com/vkpapps/thunder/ui/adapter/AudioAdapter.kt
package com.vkpapps.thunder.ui.adapter

import android.content.Context
import android.content.Intent
import android.media.MediaScannerConnection
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.RadioButton
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.widget.AppCompatImageView
import androidx.recyclerview.widget.RecyclerView
import com.vkpapps.thunder.R
import com.vkpapps.thunder.analitics.Logger
import com.vkpapps.thunder.model.AudioInfo
import com.vkpapps.thunder.ui.adapter.AudioAdapter.AudioViewHolder
import com.vkpapps.thunder.utils.IntentUtils
import com.vkpapps.thunder.utils.MyThumbnailUtils


/**
 * @author VIJAY PATIDAR
 */
class AudioAdapter(private val audioInfos: MutableList<AudioInfo>, private val onAudioSelectedListener: OnAudioSelectedListener, context: Context) : RecyclerView.Adapter<AudioViewHolder>() {
    private val inflater: LayoutInflater = LayoutInflater.from(context)
    private val myThumbnailUtils = MyThumbnailUtils

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): AudioViewHolder {
        val inflate = inflater.inflate(R.layout.audio_list_item, parent, false)
        return AudioViewHolder(inflate)
    }

    override fun onBindViewHolder(holder: AudioViewHolder, position: Int) {
        val audioinfo = audioInfos[position]

        holder.audioTitle.text = audioinfo.name
        holder.btnSelect.isChecked = audioinfo.isSelected
        holder.btnSelect.setOnClickListener {
            audioinfo.isSelected = !audioinfo.isSelected
            holder.btnSelect.isChecked = audioinfo.isSelected
            if (audioinfo.isSelected) {
                onAudioSelectedListener.onAudioSelected(audioinfo)
            } else {
                onAudioSelectedListener.onAudioDeselected(audioinfo)
            }
        }
        holder.audioIcon.setOnClickListener {
            IntentUtils.startIntentToPlayAudio(it.context, audioinfo.uri)
            try {
                val intent = Intent(Intent.ACTION_VIEW)
                MediaScannerConnection.scanFile(it.context, arrayOf(audioinfo.uri.path), null) { _, uri ->
                    run {
                        val type = it.context.contentResolver.getType(uri)
                        Logger.d("file $uri type = $type")
                        intent.setDataAndType(uri, type)
                        it.context.startActivity(intent)
                    }
                }
            } catch (e: Exception) {
                Toast.makeText(it.context, "error occurred", Toast.LENGTH_SHORT).show()
                e.printStackTrace()
            }
        }
        holder.itemView.setOnLongClickListener {
            onAudioSelectedListener.onAudioLongClickListener(audioinfo)
            true
        }
        myThumbnailUtils.loadAudioThumbnail(audioinfo.id, audioinfo.uri, holder.audioIcon)
        holder.audioIcon.adjustViewBounds = true
    }

    override fun getItemCount(): Int {
        return audioInfos.size
    }

    class AudioViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        var audioTitle: TextView = itemView.findViewById(R.id.audio_title)
        var audioIcon: AppCompatImageView = itemView.findViewById(R.id.audio_icon)
        var btnSelect: RadioButton = itemView.findViewById(R.id.btnSelect)
    }

    interface OnAudioSelectedListener {
        fun onAudioLongClickListener(audioinfo: AudioInfo)
        fun onAudioSelected(audioMode: AudioInfo)
        fun onAudioDeselected(audioinfo: AudioInfo)
    }
}